### PR TITLE
Add feature to use aws tags for finding node names

### PR DESCRIPTION
**Pull request by jsravn**
This requires setting `useKubernetesNodeTag = true` in the cloud config.

When enabled it does two things:
- Finds ec2 instance id by the tag 'kubernetes.io/nodename', looking for value
  equal to the node name.
- If tag is missing, tags the instance with the kubelet provided
  hostname (which can be overridden with --override-hostname).

So for instance, if kubelet is run with
`--overide-hostname=node-01.internal`, the node will tag itself with
`kubenernetes.io/nodename=node-01.internal`, and all queries will search
against this tag when looking up the instance.

This enables users to specify their own node names instead of forcing
usage of private-dns-name.

Authored with @dogan-sky (Dogan Narinc) and @balooo (Joseph Samuel).

### CODE REVIEW

**Comment by jsravn** [comment]
This is an alternative to #17643. It let's us use whatever node name we want. I'm not sure what the best approach is to do something like this, or what the kubernetes authors prefer, so we've made it a feature toggle in the aws cloud config.

**Comment by jsravn** [comment]
Dunno why the CLA check is failing - is it because we have two author names in the commit message? The email is verified on the cla.

**Comment by eparis** [comment]
ok to test

**Comment by eparis** [comment]
The CLA problem is stated: https://github.com/kubernetes/kubernetes/pull/20207#issuecomment-175547604
The commit was authored by someone other than the person who sent the commit in this PR. While the original commit author has signed the CLA (and so has the person who opened the PR) the original author needs to indicate that this commit should be allowed into the repo, then a human can manually override the bot...

**Comment by dogan-sky** [comment]
Hi, I can confirm that I am the author of this commit.

**Comment by balooo** [comment]
Hi, I can confirm that I am the author of the last commit.

File: pkg/cloudprovider/providers/aws/aws_test.go [hunk 1]
```diff
@@ -266,12 +266,37 @@ func contains(haystack []*string, needle string) bool {
 
 func instanceMatchesFilter(instance *ec2.Instance, filter *ec2.Filter) bool {
 	name := *filter.Name
+	glog.Infof("Checking filter %v on instance id %v", name, *instance.InstanceId)
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_test.go]: Use aws.StringValue instead of dereferencing the possibly-nil InstanceId (admittedly unlikely in this case!)

Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_test.go]: Oh ... this is tests.  Never mind!

File: pkg/cloudprovider/providers/aws/aws.go [hunk 2]
```diff
@@ -301,6 +304,10 @@ func (self *AWSCloud) AddSSHKeyToAllInstances(user string, keyData []byte) error
 }
 
 func (a *AWSCloud) CurrentNodeName(hostname string) (string, error) {
+	if a.cfg.Global.UseKubernetesNodeTag {
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Can we just use the KubernetesNodeTag here, rather than assuming that hostname == KubernetesNodeTag?  We're about to get it anyway when we call getSelfAWSInstance

File: pkg/cloudprovider/providers/aws/aws.go [hunk 3]
```diff
@@ -2191,10 +2198,19 @@ func (a *AWSCloud) getInstancesByNodeNames(nodeNames []string) ([]*ec2.Instance,
...
+		filterKey = "tag:" + TagNameKubernetesNode
+	} else {
+		filterKey = "private-dns-name"
+	}
+
 	filters := []*ec2.Filter{
-		newEc2Filter("private-dns-name", nodeName),
+		newEc2Filter(filterKey, nodeName),
+		newEc2Filter("instance-state-name", "running"),
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: I think adding this is-running filter here is probably a good thing, but it is a behavioural change.  Can we split it out into its own PR.  Or is it important to have in this PR for some reason?

File: pkg/cloudprovider/providers/aws/aws.go [hunk 4]
```diff
@@ -52,6 +52,7 @@ const ProviderName = "aws"
 
 // The tag name we use to differentiate multiple logically independent clusters running in the same AZ
 const TagNameKubernetesCluster = "KubernetesCluster"
+const TagNameKubernetesNode = "KubernetesNode"
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws.go]: Can we make this `kubernetes.io/node` or `kubernetes.io/nodename`.  I screwed up before by not prefixing KubernetesCluster, and eventually I'm going to have to fix it...

File: pkg/cloudprovider/providers/aws/aws_test.go [hunk 5]
```diff
@@ -266,12 +266,37 @@ func contains(haystack []*string, needle string) bool {
...
+
 	if name == "private-dns-name" {
 		if instance.PrivateDnsName == nil {
 			return false
 		}
 		return contains(filter.Values, *instance.PrivateDnsName)
 	}
+
+	if name == "tag:"+TagNameKubernetesNode {
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_test.go]: Can we do
```
if strings.HasPrefix(name, "tag:") {
  tagName := name[4:]
...
```

File: pkg/cloudprovider/providers/aws/aws_test.go [hunk 6]
```diff
@@ -299,8 +324,11 @@ func (self *FakeEC2) DescribeInstances(request *ec2.DescribeInstancesInput) ([]*
 			allMatch := true
 			for _, filter := range request.Filters {
 				if !instanceMatchesFilter(instance, filter) {
+					glog.Info("Filter didn't match")
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_test.go]: Debug code?  Let's remove if so.

File: pkg/cloudprovider/providers/aws/aws_test.go [hunk 7]
```diff
@@ -299,8 +324,11 @@ func (self *FakeEC2) DescribeInstances(request *ec2.DescribeInstancesInput) ([]*
...
 					allMatch = false
 					break
+				} else {
+					glog.Info("Filter matched")
```
Comment (justinsb) [inline pkg/cloudprovider/providers/aws/aws_test.go]: Debug code?  Let's remove if so.

**Comment by justinsb** [comment]
This is nice, except I don't know how you intend to add these tags.  Is the idea that you will launch your nodes manually, instead of in an ASG?  I worry that it's not very useful in practice if you have to do that.

**Comment by jsravn** [comment]
> This is nice, except I don't know how you intend to add these tags. Is the idea that you will launch your nodes manually, instead of in an ASG? I worry that it's not very useful in practice if you have to do that.

Good point. We were tagging it using some separate provisioning scripts. We could change kubelet to tag the instance it's running on, then it auto configures.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 3 above)
Comment (jsravn) [inline pkg/cloudprovider/providers/aws/aws.go]: It's more of an issue with the node labels, since there will be duplicates and it will fail.

Without node labels, just searching on 'KubernetesCluster', it will potentially attach terminated instances. I don't think this is an issue as health check will fail and it won't route traffic.

I am happy to make a separate PR though and can do that, since it is somewhat an orthogonal change.

Comment (jsravn) [inline pkg/cloudprovider/providers/aws/aws.go]: We've extracted this into #20809

File: pkg/cloudprovider/providers/aws/aws.go (hunk 2 above)
Comment (jsravn) [inline pkg/cloudprovider/providers/aws/aws.go]: We'd have to do a `DescribeTags` call I believe to get it. Conceptually it's better but it is a bit more code and an additional aws api call.

**Comment by jsravn** [comment]
Going to have to rework this a bit, as master has changed quite a bit in the aws provider. I'll do that and rebase/squash it and repush.

As per my last comment, I'm planning to change `CurrentNodeName` to tag the instance if the tag is missing.

**Comment by jsravn** [comment]
@justinsb I've reworked for latest changes, and also hopefully addressed all your prior comments.

It now tags the node at startup (when kubelet calls CurrentNodeName) so it doesn't need a separate process.

**Comment by justinsb** [comment]
We merged #20378 as it seemed a very direct fix to a regression that was actually introduced in 1.2.

Do you still think we should consider this approach too @jsravn ?  Are there additional benefits to using a tag instead of the DNS name, given that the node name must be resolvable anyway?

**Comment by jsravn** [comment]
@justinsb with #20378 this seems less necessary. The advantage of this change is flexibility in specifying node name - for us, we have our own provisioning so we were assigning human readable node names (e.g. `node-01`). But I don't think it makes much sense in the context of having an auto scaling group. I'll try out #20378 and if that works for us we can stop using this change. So I'm happy to close this PR.

**Comment by justinsb** [comment]
Thanks!  Feel free to reopen if you discover a need.
